File: C++/659.cpp
__________________________________________________________________________________________________
sample 28 ms submission
auto fast_io = []() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);
    return nullptr;
}();

class Solution {
public:
    bool isPossible(vector<int>& nums) {
        deque<int> start;
        int idx = 0, idx2 = 0, prev = 0, curr = nums[0];
        while (idx < nums.size()) {
            while (idx2+1 < nums.size() && nums[idx2+1] == curr) ++idx2;
            int cnt = idx2-idx+1;
            if (cnt > prev) {
                for (int i = prev; i < cnt; ++i) {
                    start.emplace_back(curr);
                    // cout<<"start: "<<curr<<endl;
                }
            } else if (cnt < prev) {
                for (int i = cnt; i < prev; ++i) {
                    int temp = start.front();
                    start.pop_front();
                    // cout<<"end: "<<curr<<endl;
                    if (curr-temp < 3) return false;
                }
            }
            prev = cnt;
            idx = idx2+1;
            ++curr;            
        }
        for (int i = 0; i < prev; ++i) {
            int temp = start.front();
            start.pop_front();
            // cout<<"end: "<<curr<<endl;
            if (curr-temp < 3) return false;
        }
        return true;
    }
};
__________________________________________________________________________________________________
sample 15548 kb submission
class Solution {
public:
    bool isPossible(vector<int>& nums) {
        if(nums.size()==0) return true;
        int count =0;
        int num = INT_MIN;
		vector<int> cnt;
        for(int i=0;i<=nums.size();i++){
            if(count==0){
				cnt.push_back(0);
                count=1;
                num=nums[i];
            }
            else if(i==nums.size()||nums[i]!=num){   
				cnt.push_back(count);
				if(i==nums.size()||nums[i]-num>1){
					cnt.push_back(0);
				}
                if(i==nums.size())
                    break;
                count=1;
                num=nums[i];               
            }  
            else{                
                count++;
            }
        }
		queue<int> s;
		for(int i=1;i<cnt.size();i++){
			if(cnt[i]>cnt[i-1]){
				int h = cnt[i]-cnt[i-1];
				for(int m=0;m<h;m++){
					s.push(i);
				}
			}
			else if(cnt[i]<cnt[i-1]){
				int h = cnt[i-1]-cnt[i];
				for(int m=0;m<h;m++){
					int startIndex = s.front();
					s.pop();
					if(i-startIndex<=2)return false;
				}
			}
		}
		
        return true;
    }
};
__________________________________________________________________________________________________

File: Python3/659.py
__________________________________________________________________________________________________
sample 584 ms submission
class Solution:
    def isPossible(self, nums: List[int]) -> bool:
        if len(nums) < 3:
            return False
        i = 0
        p1 = 0
        p2 = 0
        p3 = 0
        prev = -2
        while i < len(nums):
            count = 1
            while i + 1 < len(nums) and nums[i] == nums[i+1]:
                i += 1
                count += 1
            if nums[i] != prev + 1:
                if p1 > 0 or p2 > 0:
                    return False
                p1 = count
                p2 = 0
                p3 = 0
            else:
                if count < p1 + p2:
                    return False
                c1 = max(0, count - p1 - p2 - p3)
                c2 = p1
                c3 = p2 + min(p3, count - p2 - p1)
                p1 = c1 
                p2 = c2
                p3 = c3
            prev = nums[i]
            i += 1
        return True if p1 == 0 and p2 == 0 else False
__________________________________________________________________________________________________
sample 14012 kb submission
class Solution:
    def isPossible(self, nums: List[int]) -> bool:
        count = collections.Counter(nums)
        tails = collections.Counter()
        for x in nums:
            if count[x] == 0:
                continue
            elif tails[x] > 0:
                tails[x] -= 1
                tails[x+1] += 1
            elif count[x+1] > 0 and count[x+2] > 0:
                count[x+1] -= 1
                count[x+2] -= 1
                tails[x+3] += 1
            else:
                return False
            count[x] -= 1
        return True        
__________________________________________________________________________________________________
